core/lowering/passes/exception_elimination.cpp

```diff
@@ -108,6 +109,59 @@ void EliminateExceptionOrPassPattern(std::shared_ptr<Graph> graph) {
   }
 }
 
+/*
+  Below is a fork of the torch::jit::EliminateExceptions pass, with node replacement
+  using replaceAllUsesDominatedByNodeWith instead of replaceAllUsesWith,
+  so as to not invalidate the IR in challenging cases, such as nested Ifs
+
+  Original Source from which it was adapted:
+  https://github.com/pytorch/pytorch/blob/c29ab84115f40614d04e4557ea2e1ac40b7aa75c/torch/csrc/jit/passes/remove_exceptions.cpp
+*/
+
+bool certainlyThrows(Block* block) {
+  // A block certainly throws an exception if it contains
+  // the prim::RaiseException operation
+  for (Node* n : block->nodes()) {
+    if (n->kind() == prim::RaiseException) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void EliminateExceptionsSafe(Block* block) {
+  auto graph = block->owningGraph();
+  // Generate false and true constant placeholders
+  Value* false_const = graph->insertConstant(IValue(false));
+  Value* true_const = graph->insertConstant(IValue(true));
+
+  // For each prim::If node, if either block certainly throws an exception
+  // Replace all uses of the node input with the logical opposite
+  for (Node* n : block->nodes()) {
+    if (n->kind() == prim::If) {
+      Block* true_block = n->blocks()[0];
+      Block* false_block = n->blocks()[1];
+
+      if (certainlyThrows(true_block)) {
+        n->input(0)->replaceAllUsesDominatedByNodeWith(n, false_const);
+      } else if (certainlyThrows(false_block)) {
+        n->input(0)->replaceAllUsesDominatedByNodeWith(n, true_const);
```

Comment:
This does not make sense. 
If you check what's actually going on for `prim::If` [here](https://github.com/pytorch/pytorch/blob/8e0837cf84d9d2cc4996ba696ee41b5dd2d008b9/torch/csrc/jit/ir/ir_views.h), you will find that: 
1. every if node has some blocks and attributes
2. for if node, `IfNode->input(0)` is actually the condition
So, in PyTorch's pass implementation [here](https://github.com/pytorch/pytorch/blob/c29ab84115f40614d04e4557ea2e1ac40b7aa75c/torch/csrc/jit/passes/remove_exceptions.cpp#L28), what's happening is that if will go through the 2 blocks in `prim::If`, if one of that block contains a `prim::RaiseException` node, it will RESET the `prim::If` input(condition) to enforce it to run another flow. 
For example, if the original condition is `True` it should go to the first block(true block) but it finds out the first block contains a `prim::RaiseException` node, so it rewrite the condition to be `False` to make sure that it won't go to the first block to avoid the exception. 

Moreover, if you go to the definition of the function `replaceAllUsesDominatedByNodeWith` [here](https://github.com/pytorch/pytorch/blob/31f311a816c026bbfca622d6121d6a7fab44260d/torch/csrc/jit/ir/ir.h#L297), you will find that this actually replace the value dominated by the node. In graph theory, [dominate](https://en.wikipedia.org/wiki/Dominator_(graph_theory)) means that the value is dominated by the node. For example, node A dominate value x means that in that graph it must flow from node A to value x. So, for this line `        n->input(0)->replaceAllUsesDominatedByNodeWith(n, true_const);` what you are doing is that you are going to replace the `prim::If` condition value which is dominated by that `prim::If` node. However, `prim::If` condition value dominate that `prim::If` node, in other words, you are replacing nothing here because `n->input(0)` is never dominated by that `n` node. 
So I think you are basically doing nothing here. 
Above is my understanding, please correct me if I'm wrong.

Reply:
Hi @bowang007 - thank you for the detailed review and comment. I agree with your observation about the replacement of all uses of the condition dominated by `n` as effectively doing nothing here. I've run + added some additional tests to the PR and I have now updated the call to **`replaceFirstUseWith`**, which I believe is the correct replacement in this case.

Based on your comments, I now think the reason for which the original halting bug in #1823 was occurring is the following:

```python
%out1.1 : Tensor = prim::If(%71)
    block0():
      ...
      -> (%73)
    block1():
         ...
        block0():
          ...
          -> ()
        block1():
          -> ()
        ...
       = prim::RaiseException(%76, %59)
      -> (%63)
  %out2.1 : Tensor = prim::If(%71)
    block0():
      ...
      -> (%82)
    block1():
      ...
```

In the graph above, the condition `%71` appears as the predicate of two _distinct_ `prim::If` nodes. When we call **`replaceAllUsesWith`**, as `torch::jit::passes::EliminateExceptions` does, this replaces downstream `prim::If` users of the same condition, and seems to cause an [infinite loop here](https://github.com/pytorch/pytorch/blob/31f311a816c026bbfca622d6121d6a7fab44260d/torch/csrc/jit/ir/ir.cpp#LL931C25-L931C25). By replacing only the **first** use of the condition, however, this would ensure _only_ the condition of the `prim::If` being directly analyzed will be replaced, avoiding this loop. What do you think of the new changes?

Reply:
Not too much to add here, the discussion seems accurate. Here is the official documentation on block and conditional semantics in PyTorch https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/OVERVIEW.md#block